numpy/core/src/common/npy_hashtable.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * This functionality is designed specifically for the ufunc machinery to
+ * dispatch based on multiple DTypes.  Since this is designed to be used
+ * as purely a cache, it currently does no reference counting.
+ * Even though this is a cache, there is currently no maximum size.  It may
+ * make sense to limit the size, or count collisions:  If too many collisions
+ * occur, we could grow the cache, otherwise, just replace an old item that
+ * was presumably not used for a long time.
+ *
+ * If a different part of NumPy requires a custom hashtable, the code should
+ * be reused with care since specializing it more for the ufunc dispatching
+ * case is likely desired.
+ */
+
+#include "templ_common.h"
+#include "npy_hashtable.h"
+
+
+
+#if SIZEOF_PY_UHASH_T > 4
+#define _NpyHASH_XXPRIME_1 ((Py_uhash_t)11400714785074694791ULL)
+#define _NpyHASH_XXPRIME_2 ((Py_uhash_t)14029467366897019727ULL)
+#define _NpyHASH_XXPRIME_5 ((Py_uhash_t)2870177450012600261ULL)
+#define _NpyHASH_XXROTATE(x) ((x << 31) | (x >> 33))  /* Rotate left 31 bits */
+#else
+#define _NpyHASH_XXPRIME_1 ((Py_uhash_t)2654435761UL)
+#define _NpyHASH_XXPRIME_2 ((Py_uhash_t)2246822519UL)
+#define _NpyHASH_XXPRIME_5 ((Py_uhash_t)374761393UL)
+#define _NpyHASH_XXROTATE(x) ((x << 13) | (x >> 19))  /* Rotate left 13 bits */
+#endif
+
+/*
+ * This hashing function is basically the Python tuple hash with the type
+ * identity hash inlined. The tuple hash itself is a reduced version of xxHash.
+ *
+ * Users cannot control pointers, so we do not have to worry about DoS attacks?
+ */
+static NPY_INLINE Py_hash_t
+identity_list_hash(PyObject *const *v, int len)
+{
+    Py_uhash_t acc = _NpyHASH_XXPRIME_5;
+    for (int i = 0; i < len; i++) {
+        /*
+         * Lane is the single item hash, which for us is the rotated pointer.
+         * Identical to the python type hash (pointers end with 0s normally).
+         */
+        size_t y = (size_t)v[i];
+        Py_uhash_t lane = (y >> 4) | (y << (8 * SIZEOF_VOID_P - 4));
+        acc += lane * _NpyHASH_XXPRIME_2;
+        acc = _NpyHASH_XXROTATE(acc);
+        acc *= _NpyHASH_XXPRIME_1;
+    }
+    return acc;
+}
+#undef _NpyHASH_XXPRIME_1
+#undef _NpyHASH_XXPRIME_2
+#undef _NpyHASH_XXPRIME_5
+#undef _NpyHASH_XXROTATE
+
+
+static NPY_INLINE PyObject **
+find_item(PyArrayIdentityHash const *tb, PyObject *const *key)
+{
+    Py_hash_t hash = identity_list_hash(key, tb->key_len);
+    npy_uintp perturb = (npy_uintp)hash;
+    npy_intp bucket;
+    npy_intp mask = tb->size - 1 ;
+    PyObject **item;
+
+    bucket = (npy_intp)hash & mask;
+    while (1) {
```

Comment:
Like @charris, I must admit I was surprised by this! I guess one could put `perturb` here and have a `RuntimeError` at the end... Then even if someone overwrote the cache by accident this could never go into an endless loop.

Reply:
Yeah, the loop finishes due to the fact that the table is always partially empty and the perturbation strategy should visit everything eventually (apparently).  IIRC, I had a type bug here when copying from Python that caused an incorrect perturb and the windows CI actually went into the infinite loop.

So yeah, its reasonable to be worried about... at least the infinite-loop case is a "loud" failure so to speak ;).

Reply:
I suppose one could use a for loop to limit the number of iterations. OTOH, this looks like something that is probably in the standard literature somewhere. Last time I implemented a hash table it used a linear search from the collision location, which is (was) also a standard method, but probably doesn't keep entries scattered as well.